File: backend/app/schemas/portfolio.py
from datetime import datetime
from pydantic import BaseModel

from app.utils.ticker import Market, Currency, BeijingDateTime


class HoldingSummary(BaseModel):
    id: int
    symbol: str
    name: str
    market: str
    quantity: float
    cost_price: float
    holding_ratio: float
    contract_multiplier: float
    margin_rate: float
    current_price: float | None
    currency: str
    market_value: float
    market_value_base: float
    cost_total: float
    gain_loss: float
    gain_loss_pct: float
    weight_pct: float
    price_updated_at: BeijingDateTime | None
    linked_broker_holding_id: int | None = None
    broker_account_name: str | None = None


class RealizedPnlItem(BaseModel):
    holding_id: int
    symbol: str
    name: str
    currency: str
    realized_pnl_native: float
    realized_pnl_base: float
    source: str = "own"  # "own" | "linked"


class MarketBreakdown(BaseModel):
    value: float
    weight_pct: float


class PortfolioSummary(BaseModel):
    base_currency: str
    total_market_value: float
    total_cost: float
    total_gain_loss: float
    total_gain_loss_pct: float
    total_realized_pnl: float
    total_cash: float
    futures_margin: float = 0.0
    cash_balances: dict[str, float]
    holdings: list[HoldingSummary]
    realized_pnl_details: list[RealizedPnlItem] = []
    by_market: dict[str, MarketBreakdown]
    by_currency: dict[str, MarketBreakdown]
    exchange_rates: dict[str, float]
    last_refreshed: BeijingDateTime | None


class ExchangeRateResponse(BaseModel):
    from_currency: str
    to_currency: str
    rate: float
    updated_at: BeijingDateTime

    model_config = {"from_attributes": True}

File: frontend/src/types/portfolio.ts
export interface HoldingSummary {
  id: number
  symbol: string
  name: string
  market: string
  quantity: number
  cost_price: number
  holding_ratio: number
  contract_multiplier: number
  margin_rate: number
  current_price: number | null
  currency: string
  market_value: number
  market_value_base: number
  cost_total: number
  gain_loss: number
  gain_loss_pct: number
  weight_pct: number
  price_updated_at: string | null
  linked_broker_holding_id: number | null
  broker_account_name: string | null
}

export interface RealizedPnlItem {
  holding_id: number
  symbol: string
  name: string
  currency: string
  realized_pnl_native: number
  realized_pnl_base: number
  source: 'own' | 'linked'
}

export interface MarketBreakdown {
  value: number
  weight_pct: number
}

export interface PortfolioSummary {
  base_currency: string
  total_market_value: number
  total_cost: number
  total_gain_loss: number
  total_gain_loss_pct: number
  total_realized_pnl: number
  total_cash: number
  cash_balances: Record<string, number>
  holdings: HoldingSummary[]
  realized_pnl_details: RealizedPnlItem[]
  by_market: Record<string, MarketBreakdown>
  by_currency: Record<string, MarketBreakdown>
  exchange_rates: Record<string, number>
  last_refreshed: string | null
}
